File: LeetCode75_LevelTwo/Longest_Palindrome_by_Concatenating_Two_Letter_Words/Longest_Palindrome_by_Concatenating_Two_Letter_Words.cc
// 2131. Longest Palindrome by Concatenating Two Letter Words C++
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        std::unordered_map<string, int> mp;
        int ans = 0, middle = 0;
        for (auto x: words) mp[x]++;
        for (auto [s, f]: mp) {
            string re(s.rbegin(), s.rend());
            if (s == re) {
                ans += f;
                if (f % 2 == 1) {
                    middle = 1;
                    ans -= 1;
                }
            } else if (mp.find(re) != mp.end()) {
                ans += min(mp[s], mp[re]);
            }
        }
        ans += middle;
        return ans * 2;
    }
};

File: LeetCode75_LevelTwo/Longest_Palindrome_by_Concatenating_Two_Letter_Words/Longest_Palindrome_by_Concatenating_Two_Letter_Words.py
# 2131. Longest Palindrome by Concatenating Two Letter Words Python

class Solution(object):
    def longestPalindrome(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        wc = collections.Counter(words)
        iden = 0
        center = 0
        abba = 0
        for w, c in wc.items():
            if w[0] == w[1]:
                iden += c // 2 * 2
                if c % 2 == 1: center = 2
            else:
                abba += min(wc[w], wc[w[::-1]]) * 0.5
        return iden * 2 + int(abba) * 4 + center
